crates/bin/prove_block/src/main.rs

```diff
@@ -321,146 +387,113 @@ async fn main() -> Result<(), Box<dyn Error>> {
     let mut contract_storages = ContractStorageMap::new();
 
-    let nonces: HashMap<Felt252, Felt252> =
-        state_update.state_diff.nonces.iter().map(|nu| (nu.contract_address, nu.nonce)).collect();
-
-    let num_storage_diffs = state_update.state_diff.storage_diffs.len();
-    let mut updates = Vec::with_capacity(num_storage_diffs);
-    for i in 0..num_storage_diffs {
-        let storage_diff_item = &state_update.state_diff.storage_diffs[i];
-        let storage_proof = &storage_proofs[&storage_diff_item.address];
-        let contract_address = storage_diff_item.address;
-
-        // TODO: as done below, we need to use the previous nonce, not the updated one. however, there isn't
-        // really any overlap between contracts with storage updates (no nonce updates) and account
-        // contracts (nonces are updated), so this would be a corner case for now
-        let nonce = nonces.get(&contract_address).copied();
-
-        log::debug!("contract with storage diff, address: {} (nonce: {:?})", contract_address, nonce);
-        let contract_address_biguint = contract_address.to_biguint();
-
-        let address: ContractAddress = ContractAddress(PatriciaKey::try_from(felt_vm2api(contract_address)).unwrap());
-        let initial_contract_state = initial_state.get_contract_state(address)?;
-        let initial_tree = initial_contract_state.storage_commitment_tree.clone();
-
-        let contract_state = initial_contract_state
-            .update(
-                &mut initial_state.ffc,
-                &storage_diff_item.storage_entries.iter().map(|entry| (entry.key, entry.value)).collect(),
-                nonce,
-                address_to_class_hash.get(&contract_address).copied(),
-            )
-            .await?;
-        let updated_tree = contract_state.storage_commitment_tree.clone();
-
-        if storage_proof.class_commitment != contract_state.storage_commitment_tree.root.into() {
-            log::error!(
-                "expected class_commitment != computed class_commitment ({:?} != {:?})",
-                storage_proof.class_commitment,
-                contract_state.storage_commitment_tree.root.clone()
-            );
-        }
+    for (contract_address, _storage_proof) in storage_proofs {
+        let previous_storage_proof =
+            previous_storage_proofs.get(&contract_address).expect("failed to find previous storage proof");
 
-        updates.push((contract_address_biguint, contract_state.clone()));
-        contract_states.insert(storage_diff_item.address, contract_state);
+        let previous_tree = PatriciaTree { root: previous_storage_proof.state_commitment.into(), height: Height(251) };
 
-        let contract_storage =
-            OsSingleStarknetStorage::new(initial_tree, updated_tree, &[], initial_state.ffc.clone()).await?;
+        let contract_storage = ProverPerContractStorage::new(previous_block_id, contract_address, provider_url.clone());
```

Comment:
Why do we need this? Can you add a comment?
@notlesh

Reply:
I'm not sure what you mean, a comment about why we need storage for each contract?